### PR TITLE
feat(streaming): estimated cache size for hash join

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR
- introduces an `EstimateSize` trait for estimating memory usage of structs, which can be used for determining cache operations
- add a panel in grafana of "Join Cached Estimated Size"

Note that this PR does not introduce a weighted LRU cache (some attempts in #4997).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #4988

### CODE REVIEW

File: src/stream/src/executor/hash_join.rs [hunk 1]
```diff
@@ -610,14 +606,12 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
...
+        // `commit` them here.
         self.side_l.ht.flush().await?;
         self.side_r.ht.flush().await?;
 
-        // evict the LRU cache
-        // assert!(!self.side_l.is_dirty());
-        self.side_l.ht.evict_to_target_cap();
-        // assert!(!self.side_r.is_dirty());
-        self.side_r.ht.evict_to_target_cap();
+        // Note: the LRU cache is automatically evicted as we operate on it.
```
Comment (yuhao-su) [inline src/stream/src/executor/hash_join.rs:614]: Is that also true when not using moka?

File: src/common/src/array/data_chunk_iter.rs [hunk 2]
```diff
@@ -303,6 +304,13 @@ impl Row {
     }
 }
 
+impl EstimateSize for Row {
+    fn estimated_heap_size(&self) -> usize {
+        // TODO: this is not accurate now as the heap size of some `Scalar` is not counted.
```
Comment (yuhao-su) [inline src/common/src/array/data_chunk_iter.rs:309]:
```suggestion
        // FIXME(bugen): this is not accurate now as the heap size of some `Scalar` is not counted.
```

File: src/stream/src/executor/hash_join.rs (hunk 1 above)
Comment (BugenZhao) [reply src/stream/src/executor/hash_join.rs:614]: Oops. Good catch.

**Comment by yuhao-su** [comment]
Rest LGTM!

File: src/common/src/collection/estimate_size.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+/// The trait for estimating the actual memory usage of a struct.
+///
+/// Used for cache eviction now.
+pub trait EstimateSize {
+    /// The estimated heap size of the current struct in bytes.
+    fn estimated_heap_size(&self) -> usize;
+
+    /// The estimated total size of the current struct in bytes, including the `estimated_heap_size`
+    /// and the size of `Self`.
+    fn estimated_size(&self) -> usize
```
Comment (fuyufjh) [inline src/common/src/collection/estimate_size.rs:26]: - `estimated_size`
- `EstimateSize`

With or without `d`, that's a question 🥵

File: src/common/src/hash/key.rs [hunk 4]
```diff
@@ -235,11 +238,17 @@ pub struct FixedSizeKey<const N: usize> {
 /// See [`crate::hash::calc_hash_key_kind`]
 #[derive(Clone, Debug)]
 pub struct SerializedKey {
-    key: Vec<Datum>,
+    key: Row,
```
Comment (fuyufjh) [inline src/common/src/hash/key.rs:241]: (off the topic of this PR) Why was the struct named `SerializedKey`? It's actually `Vec<Datum>`...  @liurenjie1024

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/common/src/collection/estimate_size.rs (hunk 3 above)
Comment (BugenZhao) [reply src/common/src/collection/estimate_size.rs:26]: We may treat the trait name as an "action" and the method a "property". 🤣

File: src/common/src/hash/key.rs (hunk 4 above)
Comment (liurenjie1024) [reply src/common/src/hash/key.rs:241]: It was meant to serialize keys using some serialized format, and `Vec<Datum>` was just a temporary solution.

**Pull request merged**